device/transport/mqtt/src/mqtt.ts

```diff
@@ -547,4 +608,222 @@ export class Mqtt extends EventEmitter implements Client.Transport, StableConnec
+            msg.correlationId = v;
+            break;
+          case '$.uid':
+            /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_012: [When a message is received, the receiver shall populate the generated `Message` object `userId` with the value of the property `$.uid` serialized in the topic, if present.]*/
+            msg.userId = v;
+            break;
+          default:
+            /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_007: [When a message is received, the receiver shall populate the generated `Message` object `properties` property with the user properties serialized in the topic.]*/
+            msg.properties.add(k, v);
+            break;
+        }
+      }
+    }
+
+    /*Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_004: [If there is a listener for the message event, a message event shall be emitted for each message received.]*/
+    this.emit('message', msg);
+  }
+
+  private _onDeviceMethod(topic: string, payload: any): void {
+    // The topic name looks like this:
+    //  $iothub/methods/POST/{method name}?$rid={request id}&{serialized properties}
+    // We parse out the message.
+
+    /* Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_13_005: [ When a method_<METHOD NAME> event is emitted the parameter shall conform to the shape as defined by the interface specified below:
+
+      interface StringMap {
+          [key: string]: string;
+      }
+
+      interface MethodMessage {
+          methods: { methodName: string; };
+          requestId: string;
+          properties: StringMap;
+          body: Buffer;
+          verb: string;
+      }
+    ]*/
+    const methodMessage = _parseMessage(topic, payload);
+    if (!!methodMessage) {
+      // Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_13_003: [ If there is a listener for the method event, a method_<METHOD NAME> event shall be emitted for each message received. ]
+      // we emit a message for the event 'method_{method name}'
+      this.emit('method_' + methodMessage.methods.methodName, methodMessage);
+    }
+  }
+}
+
+/**
+ * @private
+ */
+interface TopicDescription {
+  name: string;
+  listenersCount: number;
+  subscribeInProgress: boolean;
+  subscribed: boolean;
+  topicMatchRegex: RegExp;
+  handler: Function;
+}
+
+/**
+ * @private
+ */
+class MethodDescription {
+  methodName: string;
+  verb: string;
+}
+
+/**
+ * @private
+ */
+class MethodMessage {
+  methods: MethodDescription;
+  requestId: string;
+  properties: { [key: string]: string };
+  body: Buffer;
+}
+
+function _parseMessage(topic: string, body: any): MethodMessage {
+  let url: URL.Url;
+  let path: string[];
+  let query: any;
+  try {
+    url = URL.parse(topic);
+    path = url.path.split('/');
+    query = querystring.parse(url.query);
+  } catch (err) {
+    return undefined;
```

Comment:
>  return undefined; [](start = 2, length = 19)

So, in _onDeviceMethod, things just silently fail if there is a problem HERE?

Reply:
if we receive something that can't be parsed, we swallow the exception (which would otherwise crash the program) and then return undefined. the return value of _parseMessage is checked by the caller, and if it is not undefined, an event to trigger the method handler is emitted. if it is undefined, nothing is emitted. Consider the fact that the server could be sending us garbage and we wouldn't want to crash. nevertheless we should log it so I added a log line here.

---
In reply to: [141921549](https://github.com/Azure/azure-iot-sdk-node/pull/96#discussion_r141921549) [](ancestors = 141921549)